examples/training/textual_inversion.py

```diff
@@ -0,0 +1,787 @@
+                if args.embedding_reg_weight > 0:
+                    num_embeddings = 1  # TODO: generalize to multiple embeddings
+                    optimized = model.text_encoder.get_concept_embeddings()
+                    coarse = model.text_encoder.get_initializer_embeddings().clone().to(optimized.device)
+                    coarse_loss = (optimized - coarse) @ (optimized - coarse).T / num_embeddings
+
+                    loss += args.embedding_reg_weight * coarse_loss
+
+                accelerator.backward(loss)
+                # accelerator.clip_grad_norm_(model.parameters(), 1.0)
+                optimizer.step()
+                lr_scheduler.step()
+                optimizer.zero_grad()
+
+            # Checks if the accelerator has performed an optimization step behind the scenes
+            if accelerator.sync_gradients:
+                progress_bar.update(1)
+                global_step += 1
+
+            logs = {"loss": loss.detach().item(), "lr": lr_scheduler.get_last_lr()[0], "step": global_step}
+            progress_bar.set_postfix(**logs)
+            accelerator.log(logs, step=global_step)
+
+            if global_step >= args.max_train_steps:
+                break
+        progress_bar.close()
+
+        accelerator.wait_for_everyone()
+
+        # Generate sample images for visual inspection
+        # if accelerator.is_main_process:
+        #     if epoch % args.save_images_epochs == 0 or epoch == args.num_epochs - 1:
+        #         pipeline = StableDiffusionPipeline(
+        #             unet=accelerator.unwrap_model(stable_diffusion.unet),
+        #             vae=accelerator.unwrap_model(stable_diffusion.vae),
+        #             text_encoder=accelerator.unwrap_model(stable_diffusion.text_encoder),
+        #             tokenizer=tokenizer,
+        #             scheduler=noise_scheduler,
+        #             safety_checker=StableDiffusionSafetyChecker.from_pretrained("CompVis/stable-diffusion-safety-checker"),
+        #             feature_extractor=CLIPFeatureExtractor.from_pretrained("openai/clip-vit-base-patch32")
+        #         )
+
+        #         generator = torch.manual_seed(0)
+        #         # run pipeline in inference (sample random noise and denoise)
+        #         images = pipeline(generator=generator, batch_size=args.eval_batch_size, output_type="numpy")["sample"]
+
+        #         # denormalize the images and save to tensorboard
+        #         images_processed = (images * 255).round().astype("uint8")
+        #         accelerator.trackers[0].writer.add_images(
+        #             "test_samples", images_processed.transpose(0, 3, 1, 2), epoch
+        #         )
+
+        #     if epoch % args.save_model_epochs == 0 or epoch == args.num_epochs - 1:
+        #         # save the model
+        #         if args.push_to_hub:
+        #             push_to_hub(args, pipeline, repo, commit_message=f"Epoch {epoch}", blocking=False)
```

Comment:
Should we implement intermidiate saving and logging. The training doesn't take long so not needed IMO.

Reply:
+1, and the script is already complex enough even without this part :D